The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9314837987/job/25639811941) [(plain)](https://github.com/rust-lang/rust/commit/1ae82d7f990ef49178306bb389faf8a212b441fc/checks/25639811941/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:a5ac7e51b41094c92402da3b24376905380afc29)
Download action repository 'actions/setup-python@v5' (SHA:82c7e631bb3cdc910f68e0081d67478d79c6982d)
Download action repository 'actions/upload-artifact@v4' (SHA:65462800fd760344b1a7b4382951275a0abb4808)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#12 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#12 DONE 0.0s

#13 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#13 0.422   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#13 0.438 Collecting boolean-py==4.0
#13 0.444   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#13 0.460 Collecting chardet==5.1.0
---
#13 3.543 Building wheels for collected packages: reuse
#13 3.544   Building wheel for reuse (pyproject.toml): started
#13 3.860   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.861   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#13 3.861   Stored in directory: /tmp/pip-ephem-wheel-cache-hbwafndz/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#13 3.863 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#13 3.885   Attempting uninstall: setuptools
#13 3.885     Found existing installation: setuptools 59.6.0
#13 3.886     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.886     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#13 3.887     Can't uninstall 'setuptools'. No files were found to uninstall.
#13 4.538 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#13 4.538 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 5.046 Collecting virtualenv
#13 5.099   Downloading virtualenv-20.26.2-py3-none-any.whl (3.9 MB)
#13 5.269      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.9/3.9 MB 23.2 MB/s eta 0:00:00
#13 5.320 Collecting platformdirs<5,>=3.9.1
#13 5.327   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 5.361 Collecting filelock<4,>=3.12.2
#13 5.368   Downloading filelock-3.14.0-py3-none-any.whl (12 kB)
#13 5.388 Collecting distlib<1,>=0.3.7
#13 5.394   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 5.404      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 54.5 MB/s eta 0:00:00
#13 5.488 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 5.649 Successfully installed distlib-0.3.8 filelock-3.14.0 platformdirs-4.2.2 virtualenv-20.26.2
#13 DONE 5.7s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      214976 kB
DirectMap2M:     7124992 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished `dev` profile [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/366da30d554719e7490950aeea5d3f3a5927e313/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-366da30d554719e7490950aeea5d3f3a5927e313-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
    Finished `release` profile [optimized] target(s) in 24.92s
##[endgroup]
fmt check
fmt: checked 3149 files
##[error]Diff in /checkout/library/std/src/thread/mod.rs at line 561:
         let main = Box::new(main);
         // SAFETY: dynamic size and alignment of the Box remain the same. See below for why the
         // lifetime change is justified.
-        let main = unsafe { Box::from_raw(Box::into_raw(main) as *mut (dyn FnOnce() + Send + 'static)) };
+        let main =
+            unsafe { Box::from_raw(Box::into_raw(main) as *mut (dyn FnOnce() + Send + 'static)) };
         Ok(JoinInner {
             // SAFETY:
             // SAFETY:
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/library/std/src/sys/sync/mutex/no_threads.rs" "/checkout/library/std/src/sys/sync/mutex/mod.rs" "/checkout/library/std/src/sys/sync/mutex/pthread.rs" "/checkout/library/std/src/sys/sync/mutex/futex.rs" "/checkout/library/std/src/sys/sync/mutex/xous.rs" "/checkout/library/std/src/sys/sync/mutex/fuchsia.rs" "/checkout/library/std/src/sys/sync/mutex/sgx.rs" "/checkout/library/std/src/sys/sync/mutex/itron.rs" "/checkout/library/std/src/sys/sync/mutex/windows7.rs" "/checkout/library/core/src/alloc/mod.rs" "/checkout/library/core/src/alloc/global.rs" "/checkout/library/core/src/alloc/layout.rs" "/checkout/library/core/src/primitive_docs.rs" "/checkout/library/std/src/sys/sync/thread_parking/mod.rs" "/checkout/library/std/src/sys/sync/thread_parking/pthread.rs" "/checkout/library/std/src/sys/sync/thread_parking/windows.rs" "/checkout/library/std/src/sys/sync/thread_parking/futex.rs" "/checkout/library/std/src/sys/sync/thread_parking/xous.rs" "/checkout/library/std/src/sys/sync/thread_parking/unsupported.rs" "/checkout/library/std/src/sys/sync/thread_parking/darwin.rs" "/checkout/library/std/src/sys/sync/thread_parking/id.rs" "/checkout/library/std/src/path/tests.rs" "/checkout/library/core/src/slice/sort.rs" "/checkout/library/core/src/slice/ascii.rs" "/checkout/library/core/src/slice/mod.rs" "/checkout/library/core/src/slice/raw.rs" "/checkout/library/core/src/slice/rotate.rs" "/checkout/library/core/src/slice/select.rs" "/checkout/library/core/src/slice/iter.rs" "/checkout/library/core/src/slice/cmp.rs" "/checkout/library/std/src/ffi/os_str/tests.rs" "/checkout/library/std/src/ffi/mod.rs" "/checkout/library/std/src/ffi/c_str.rs" "/checkout/library/std/src/ffi/os_str.rs" "/checkout/library/core/src/slice/iter/macros.rs" "/checkout/library/core/src/slice/specialize.rs" "/checkout/library/core/src/slice/index.rs" "/checkout/library/core/src/slice/memchr.rs" "/checkout/library/core/src/intrinsics.rs" "/checkout/library/core/src/ub_checks.rs" "/checkout/library/std/src/thread/mod.rs" "/checkout/library/std/src/thread/scoped.rs" "/checkout/library/std/src/thread/local/dynamic_tests.rs" "/checkout/library/std/src/thread/local/tests.rs" "/checkout/library/std/src/thread/tests.rs" "/checkout/library/std/src/thread/local.rs" "/checkout/library/std/src/lib.rs" "/checkout/library/std/src/ascii.rs" "/checkout/library/std/src/error/tests.rs" "/checkout/library/std/src/fs/tests.rs" "/checkout/library/std/src/lib.miri.rs" "/checkout/library/core/src/ffi/mod.rs" "/checkout/library/core/src/ffi/c_str.rs" "/checkout/library/core/src/lib.rs" "/checkout/library/core/src/ascii.rs" "/checkout/library/core/src/internal_macros.rs" "/checkout/library/std/src/process/tests.rs" "/checkout/library/std/src/num.rs" "/checkout/library/std/src/macros.rs" "/checkout/library/std/src/process.rs" "/checkout/library/std/src/f64.rs" "/checkout/library/std/src/pat.rs" "/checkout/library/std/src/f16/tests.rs" "/checkout/library/core/src/option.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Fri May 31 07:59:26 UTC 2024
  network time: Fri, 31 May 2024 07:59:26 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/library/std/src/sys/pal/sgx/thread.rs
+++ b/library/std/src/sys/pal/sgx/thread.rs
@@ -28,14 +28,14 @@ mod task_queue {
     }
 
     pub(super) struct Task {
-        p: Box<dyn FnOnce()>,
+        p: Box<dyn FnOnce() + Send>,
         done: JoinNotifier,
     }
 
     unsafe impl Send for Task {}
 
     impl Task {
-        pub(super) fn new(p: Box<dyn FnOnce()>) -> (Task, JoinHandle) {
+        pub(super) fn new(p: Box<dyn FnOnce() + Send>) -> (Task, JoinHandle) {
             let (done, recv) = wait_notify::new();
             let done = JoinNotifier(Some(done));
             (Task { p, done }, recv)
@@ -97,7 +97,7 @@ pub mod wait_notify {
 
 impl Thread {
     // unsafe: see thread::Builder::spawn_unchecked for safety requirements
-    pub unsafe fn new(_stack: usize, p: Box<dyn FnOnce()>) -> io::Result<Thread> {
+    pub unsafe fn new(_stack: usize, p: Box<dyn FnOnce() + Send>) -> io::Result<Thread> {
         let mut queue_lock = task_queue::lock();
         unsafe { usercalls::launch_thread()? };
         let (task, handle) = task_queue::Task::new(p);

--- a/library/std/src/sys/pal/sgx/thread.rs
+++ b/library/std/src/sys/pal/sgx/thread.rs
@@ -32,8 +32,6 @@ mod task_queue {
         done: JoinNotifier,
     }
 
-    unsafe impl Send for Task {}
-
     impl Task {
         pub(super) fn new(p: Box<dyn FnOnce() + Send>) -> (Task, JoinHandle) {
             let (done, recv) = wait_notify::new();

--- a/library/std/src/thread/mod.rs
+++ b/library/std/src/thread/mod.rs
@@ -561,7 +561,8 @@ impl Builder {
         let main = Box::new(main);
         // SAFETY: dynamic size and alignment of the Box remain the same. See below for why the
         // lifetime change is justified.
-        let main = unsafe { Box::from_raw(Box::into_raw(main) as *mut (dyn FnOnce() + 'static)) };
+        let main =
+            unsafe { Box::from_raw(Box::into_raw(main) as *mut (dyn FnOnce() + Send + 'static)) };
 
         Ok(JoinInner {
             // SAFETY:
@@ -1544,7 +1545,7 @@ struct Packet<'scope, T> {
 // The type `T` should already always be Send (otherwise the thread could not
 // have been created) and the Packet is Sync because all access to the
 // `UnsafeCell` synchronized (by the `join()` boundary), and `ScopeData` is Sync.
-unsafe impl<'scope, T: Sync> Sync for Packet<'scope, T> {}
+unsafe impl<'scope, T: Send> Sync for Packet<'scope, T> {}
 
 impl<'scope, T> Drop for Packet<'scope, T> {
     fn drop(&mut self) {

